### ISSUE TITLE
[Build] ‘struct onnxruntime::ProviderHostCPU’ has no member named ‘UpsampleBase__AdjustOutputSizeAsPolicy’ when CONTRIB ops are disabled.

### ISSUE THREAD

**Issue by MiroPsota** [microsoft/onnxruntime#21204]
### Describe the issue

Error as described, even with different library versions.
CPU only builds fine.
CUDA from https://developer.download.nvidia.com/compute/cuda/11.8.0/local_installers/cuda_11.8.0_520.61.05_linux.run
CUDNN from
https://developer.nvidia.com/downloads/compute/cudnn/secure/8.9.2/local_installers/11.x/cudnn-linux-x86_64-8.9.2.26_cuda11-archive.tar.xz/

### Target platform

Ubuntu 24.04

### Build script

```
git clone -b v1.18.1 --depth 1 https://github.com/microsoft/onnxruntime.git

export CC=/usr/bin/gcc-11
export CXX=/usr/bin/g++-11

cd onnxruntime
./build.sh \
  --config Release \
  --build_shared_lib \
  --parallel \
  --compile_no_warning_as_error \
  --skip_submodule_sync \
  --use_cuda \
  --cuda_home <path>/cuda-11.8.0 \
  --cudnn_home <path>/cudnn-linux-x86_64-8.9.2.26_cuda11-archive \
  --cmake_extra_defines \
    CMAKE_CC_COMPILER=/usr/bin/gcc-11 \
    CMAKE_CXX_COMPILER=/usr/bin/g++-11 \
    CMAKE_CUDA_ARCHITECTURES=86 \
    CMAKE_CUDA_HOST_COMPILER=/usr/bin/g++-11 \
    onnxruntime_DISABLE_CONTRIB_OPS=ON \
    onnxruntime_BUILD_UNIT_TESTS=OFF \
  --skip_tests
```

### Error / output

<path>/onnxruntime/onnxruntime/core/providers/shared_library/provider_bridge_provider.cc:580:14: error: ‘struct onnxruntime::ProviderHostCPU’ has no member named ‘UpsampleBase__AdjustOutputSizeAsPolicy’

### GCC / Compiler Version

11.4

**Comment by xadupre** [comment]
Thanks for reporting this issue, it seems method AdjustOutputSizeAsPolicy is sometimes excluded if DISABLE_CONTRIB_OPS is on, sometimes not like here https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/shared_library/provider_bridge_provider.cc#L586

**Comment by MiroPsota** [comment]
It indeed builds without `onnxruntime_DISABLE_CONTRIB_OPS=ON`. Any tips how to fix it for `1.18.1`?

**Comment by snnn** [comment]
I submitted a PR for this.

**Comment by MiroPsota** [comment]
I tried this fix already, but then I could not create `CUDAExecutionProvider`:
```
[E:onnxruntime:Default, provider_bridge_ort.cc:1745 TryGetProviderInfo_CUDA] <redacted>/onnxruntime-1.18.1/onnxruntime/core/session/provider_bridge_ort.cc:1426 onnxruntime::Provider& onnxruntime::ProviderLibrary::Get() [ONNXRuntimeError] : 1 : FAIL : Failed to load library libonnxruntime_providers_cuda.so with error: <redacted>/python3.11/site-packages/onnxruntime/capi/libonnxruntime_providers_cuda.so: undefined symbol: _ZN11onnxruntime4cuda21BuildKernelCreateInfoINS0_57kCudaExecutionProvider_GridSample_kOnnxDomain_ver16_floatEEENS_16KernelCreateInfoEv
```
which is after demangling
```
onnxruntime::KernelCreateInfo onnxruntime::cuda::BuildKernelCreateInfo<onnxruntime::cuda::kCudaExecutionProvider_GridSample_kOnnxDomain_ver16_float>()
```

**Comment by snnn** [comment]
A side question: may I know why do you need to set onnxruntime_DISABLE_CONTRIB_OPS to ON? The build option is for reducing binary size. However, CUDA itself is already very large.  Disabling contrib ops does not help much.

**Comment by snnn** [comment]
I will update the PR.

**Comment by MiroPsota** [comment]
I like the smaller size, but I also appreciate the much faster compilation. I think it's more than 2-3 times faster on my PC, and I'm currently compiling for just one CUDA architecture. This difference includes the fact that I can use more threads, as compiling contrib kernels takes up too much RAM.

**Comment by snnn** [comment]
Would you mind trying it again? Now it should be good.

**Comment by MiroPsota** [comment]
I applied the patch to version 1.18.1 and it works.

**Comment by snnn** [comment]
Open onnxruntime/core/providers/tensorrt/onnx_ctx_model_helper.cc, find line 127, replace ORT_ENFORCE with ORT_THROW_IF_ERROR.

The new code should be:
```
ORT_THROW_IF_ERROR(graph_build.Resolve());
```

Then please let us know what error message you see.

**Comment by MiroPsota** [comment]
```
[E:onnxruntime:, inference_session.cc:2044 operator()] Exception during initialization: <path>/onnxruntime/core/providers/tensorrt/onnx_ctx_model_helper.cc:127 onnx::ModelProto* onnxruntime::CreateCtxModel(const onnxruntime::GraphViewer&, std::string, char*, size_t, int64_t, std::string, std::string, const onnxruntime::logging::Logger*) [ONNXRuntimeError] : 1 : FAIL : Fatal error: com.microsoft:EPContext(-1) is not a registered function/op
```

**Comment by snnn** [comment]
Ok .. As it says , it is not supported. You may open a new issue for this error.

**Comment by snnn** [comment]
The "com.microsoft:EPContext" op is a contrib op, for implementing TensorRT EP.

**Comment by MiroPsota** [comment]
The interesting thing is that it will create an .engine file and it works with subsequent runs, until I delete the .engine file.

**Comment by snnn** [comment]
The issue has been fixed in #20695

**Comment by snnn** [comment]
Now the build passes, but the built binary still won't work well since our CUDA execution provider needs to support both NCHW and NHWC. One of them is implemented through contrib ops. So when contrib ops are disabled, even basic ops like Conv won't work well.  Therefore I think we should disable this feature instead.
